javascript/ql/src/semmle/javascript/frameworks/jQuery.qll

```diff
@@ -258,26 +258,99 @@ private class JQueryChainedElement extends DOM::Element, InvokeExpr {
 }
 
 /**
- * A model of a URL request made using the `jQuery.ajax` or `jQuery.getJSON`.
+ * A model of a URL request made using the `jQuery.ajax`.
  */
-private class JQueryClientRequest extends ClientRequest::Range {
-  JQueryClientRequest() {
-    exists(string name |
-      name = "ajax" or
-      name = "getJSON"
-    |
-      this = jquery().getAMemberCall(name)
-    )
-  }
+private class JQueryAjaxCall extends ClientRequest::Range {
+  JQueryAjaxCall() { this = jquery().getAMemberCall("ajax") }
 
   override DataFlow::Node getUrl() {
-    result = getArgument(0) or
+    result = getArgument(0) and not exists(getOptionArgument(0, _)) or
     result = getOptionArgument([0 .. 1], "url")
   }
 
   override DataFlow::Node getHost() { none() }
 
   override DataFlow::Node getADataNode() { result = getOptionArgument([0 .. 1], "data") }
+
+  private string getResponseType() {
+    getOptionArgument([0 .. 1], "dataType").mayHaveStringValue(result)
+  }
+
+  override DataFlow::Node getAResponseDataNode(string responseType, boolean promise) {
+    (
+      responseType = getResponseType()
+      or
+      not exists(getResponseType()) and responseType = ""
+    ) and
+    promise = false and
+    result =
+      getOptionArgument([0 .. 1], "success")
```

Comment:
the `error` and `complete` handlers should be added here as well.

Reply:
I don't think they should. 

The `getAResponseDataNode` predicate returns a `Node` that refers to some representation of the response. 
The callbacks to both the `error` and the `complete` are given a `jqXHR` as the first argument, and a `String textStatus` as the second.  
None of those arguments are really a response. 

I added support for `.done` and `.then` callbacks. E.g. `$.get(url).done((data) => {..})`.

Reply:
OK, I thought the interface was more uniform. Lets skip that then.
We could probably use some property of the `jqXHR` parameter for `complete` and `error` as the response data node, but that may in the area of diminishing returns.

Reply:
> but that may in the area of diminishing returns.  

I'll do a small exploratory query on LGTM, and implement something if I get results.

Reply:
> I'll do a small exploratory query on LGTM, and implement something if I get results.

Looks like `jqXHR` is used enough to warrant a model. https://lgtm.com/query/4017763656551171517/  
I'm surprised.